File: Java/src/searches/uninformed/BreadthFirstSearch.java
package searches.uninformed;

import graphs.DirectedGraphNode;
import stack.Queue;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Chris Keyser on 3/18/2016.
 * BFS implementation for a Directed Graph (note: not just DAGs, though it will also work on a DAG)
 */
public class BreadthFirstSearch implements GraphSearch {
    /**
     * determines if graph starting at root contains target using BFS
     * @param trace if true, trace of search is printed to stdout
     */
    @Override
    public boolean contains(int target, DirectedGraphNode root, boolean trace) {
        //see if root contains target
        if(root.getVal() == target) {
            return true;
        }

        //print trace if necessary
        if(trace) {
            System.out.print(root.getVal() + " ");
        }

        //initialize queue to root's neighbors
        Queue<DirectedGraphNode> queue = new Queue<>(root.getNeighbors());
        List<DirectedGraphNode> visited = new ArrayList<>();

        //while there's still nodes to check keep looking
        while(!queue.isEmpty()) {
            //get node from queue and add to visited
            DirectedGraphNode next = queue.pop();
            visited.add(next);

            //trace if necessary
            if(trace) {
                System.out.print(next.getVal() + " ");
            }

            //check to see if this node contains target
            if(next.getVal() == target) {
                return true;
            }

            //add unvisited neighbors
            for(DirectedGraphNode node : next.getNeighbors()) {
                if(!(visited.contains(node) || queue.contains(node))) {
                    queue.push(node);
                }
            }
        }

        return false;
    }

    /**
     * returns a string representing the BFS traversal of a graph from root
     */
    @Override
    public String traverse(DirectedGraphNode root) {
        String output = root.getVal() + " ";

        Queue<DirectedGraphNode> queue = new Queue<>(root.getNeighbors());
        List<DirectedGraphNode> visited = new ArrayList<>();

        while(!queue.isEmpty()) {
            DirectedGraphNode next = queue.pop();
            visited.add(next);

            output += next.getVal() + " ";

            //add unvisited neighbors
            for(DirectedGraphNode node : next.getNeighbors()) {
                if(!(visited.contains(node) || queue.contains(node))) {
                    queue.push(node);
                }
            }
        }

        return output.trim();
    }
}

File: C#/DataStructures/DataStructures/search/graph/BreadthFirstSearch.cs
using System.Collections.Generic;
using System.Linq;

namespace DataStructures.search.graph
{
    public class BreadthFirstSearch
    {
        public static string Traverse<T>(DirectedGraphNode<T> root)
        {
            string output = "";

            var visited = new List<DirectedGraphNode<T>>();
            var queue = new Queue<DirectedGraphNode<T>>();
            queue.Enqueue(root);

            while (queue.Count != 0)
            {
                var curr = queue.Dequeue();

                foreach (var neighbor in curr.GetNeighbors())
                    if (!(visited.Any(node => node.Equals(neighbor)) || queue.Any(node => node.Equals(neighbor))))
                        queue.Enqueue(neighbor);

                visited.Add(curr);

                output += curr.Data.ToString() + " ";
            }

            return output.Trim();
        }
    }
}
